pkg/patterns/addon/pkg/status/aggregate.go

```diff
@@ -46,50 +47,90 @@ type aggregator struct {
 	log := log.Log
 
-	instance, ok := src.(addonv1alpha1.CommonObject)
-	if !ok {
+	unstruct, ok := src.(*unstructured.Unstructured)
+	instance, commonOkay := src.(addonv1alpha1.CommonObject)
+
+	unstructStatus := make(map[string]interface{})
+	var status addonv1alpha1.CommonStatus
+
+	statusHealthy := true
+	statusErrors := []string{}
+
+	if ok{
+		unstructStatus["Healthy"] = true
+	} else if commonOkay {
+		status = addonv1alpha1.CommonStatus{Healthy: true}
+	} else {
 		return fmt.Errorf("object %T was not an addonv1alpha1.CommonObject", src)
 	}
 
-	status := addonv1alpha1.CommonStatus{Healthy: true}
-
 	for _, o := range objs.Items {
 		gk := o.Group + "/" + o.Kind
 		healthy := true
 		var err error
 		switch gk {
 		case "/Service":
-			healthy, err = a.service(ctx, instance, o.Name)
+			healthy, err = a.service(ctx, src, o.Name)
 		case "extensions/Deployment", "apps/Deployment":
-			healthy, err = a.deployment(ctx, instance, o.Name)
+			healthy, err = a.deployment(ctx, src, o.Name)
 		default:
 			log.WithValues("type", gk).V(2).Info("type not implemented for status aggregation, skipping")
 		}
 
-		status.Healthy = status.Healthy && healthy
+
+		statusHealthy = statusHealthy && healthy
 		if err != nil {
-			status.Errors = append(status.Errors, fmt.Sprintf("%v", err))
+			statusErrors = append(statusErrors, fmt.Sprintf("%v", err))
 		}
 	}
 
 	log.WithValues("object", src).WithValues("status", status).V(2).Info("built status")
 
-	if !reflect.DeepEqual(status, instance.GetCommonStatus()) {
-		instance.SetCommonStatus(status)
+	if commonOkay {
+		status.Errors = statusErrors
+		status.Healthy = statusHealthy
+
+		if !reflect.DeepEqual(status, instance.GetCommonStatus()) {
+			instance.SetCommonStatus(status)
 
-		log.WithValues("name", instance.GetName()).WithValues("status", status).Info("updating status")
+			log.WithValues("name", instance.GetName()).WithValues("status", status).Info("updating status")
 
-		err := a.client.Update(ctx, instance)
+			err := a.client.Update(ctx, instance)
+			if err != nil {
+				log.Error(err, "updating status")
+				return err
+			}
+		}
+	} else {
+		unstructStatus["Healthy"] = true
+		unstructStatus["Errors"] = statusErrors
+		s, _, err := unstructured.NestedMap(unstruct.Object, "status")
 		if err != nil {
-			log.Error(err, "updating status")
-			return err
+			log.Error(err, "getting status")
+			return fmt.Errorf("unable to get status from unstructured: %v", err)
+		}
+		if !reflect.DeepEqual(status, s) {
+			err = unstructured.SetNestedField(unstruct.Object, statusHealthy, "status", "healthy")
+			err = unstructured.SetNestedStringSlice(unstruct.Object, statusErrors, "status", "errors")
+			if err != nil {
+				log.Error(err, "updating status")
+				return fmt.Errorf("unable to set status in unstructured", err)
+			}
+
+			log.WithValues("name", unstruct.GetName()).WithValues("status", status).Info("updating status")
+
+			err = a.client.Update(ctx, unstruct)
```

Comment:
If you make this `a.client.Update(ctx, src)`, can you eliminate the duplication between the two if branches?

Reply:
I don't know if this would work as the status wasn't changed on `src`. Unlike the `SetNestedField` in `unstruct` and `SetCommonStatus` in `instance` that changes the status

Reply:
Do we know where we don't change `src`?  `src` and `unstruct` are just a pointer-cast, so they should be the same object.

(Also, if we're changing status we might have to use `client.Status().Update(...)`

Reply:
It seems every time we run the update function, `src` has changed.  I will change to `client.Status().Update(...)` in another PR